kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java

```diff
@@ -0,0 +1,408 @@
+              consumer.seekToBeginning(e.partitions().toArray(new TopicPartition[0]));
+            } else {
+              consumer.seekToEnd(e.partitions().toArray(new TopicPartition[0]));
+            }
+          } catch (InterruptedException e) {
+            throw new IllegalStateException("Consumer thread is interrupted unexpectedly", e);
+          }
+        }
+      } catch (WakeupException we) {
+        logger.info("The consumer is being stopped");
+      } finally {
+        consumer.close();
+      }
+    }
+  }
+
+
+  /**
+   * This method is called in setup method of Abstract Kafka Input Operator
+   */
+  public void create(AbstractKafkaInputOperator ownerOperator)
+  {
+    holdingBuffer = new ArrayBlockingQueue<>(cacheSize);
+    this.ownerOperator = ownerOperator;
+    logger.info("Create consumer wrapper with holding buffer size: {} ", cacheSize);
+    if (logger.isInfoEnabled()) {
+      logger.info("Assignments are {} ", Joiner.on('\n').join(ownerOperator.getAssignment()));
+    }
+  }
+
+
+  /**
+   * This method is called in the activate method of the operator
+   */
+  public void start()
+  {
+    isAlive = true;
+
+
+    // thread to consume the kafka data
+    // create thread pool for consumer threads
+    kafkaConsumerExecutor = Executors.newCachedThreadPool(
+      new ThreadFactoryBuilder().setNameFormat("kafka-consumer-%d").build());
+
+    // group list of PartitionMeta by cluster
+    Map<String, List<TopicPartition>> consumerAssignment = new HashMap<>();
+    for (AbstractKafkaPartitioner.PartitionMeta partitionMeta : ownerOperator.getAssignment()) {
+      String cluster = partitionMeta.getCluster();
+      List<TopicPartition> cAssignment = consumerAssignment.get(cluster);
+      if (cAssignment == null) {
+        cAssignment = new LinkedList<>();
+        consumerAssignment.put(cluster, cAssignment);
+      }
+      cAssignment.add(new TopicPartition(partitionMeta.getTopic(), partitionMeta.getPartitionId()));
+    }
+
+    Map<AbstractKafkaPartitioner.PartitionMeta, Long> currentOffset = ownerOperator.getOffsetTrack();
+
+
+    //  create one thread for each cluster
+    // each thread use one KafkaConsumer to consume from 1+ partition(s) of 1+ topic(s)
+    for (Map.Entry<String, List<TopicPartition>> e : consumerAssignment.entrySet()) {
+
+      Properties prop = new Properties();
+      if (ownerOperator.getConsumerProps() != null) {
+        prop.putAll(ownerOperator.getConsumerProps());
+      }
+
+      prop.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, e.getKey());
+      prop.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "none");
+      // never auto commit the offsets
+      prop.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+      if (ownerOperator.getKeyDeserializer() != null) {
+        prop.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ownerOperator.getKeyDeserializer().getCanonicalName());
+      }
+      if (ownerOperator.getValueDeserializer() != null) {
```

Comment:
We can have the consumer metrics from the consumer API. But if we want to control the bandwidth bytes/window ex., there is no explicit way to do that. One thing I can think of is we always use ByteArrayDeserializer and use the actual Deserializer to deserialize the message in operator code, but I don't know if it's a right idea.